Replace `libion` with `libdmabufheap`

MarijnS95 opened the pull request:
Since kernel 5.10 ION has finally been replaced with its DMA-BUF heap successor, applying lessons learned to the kernel API.  `libion` was already a compatibility library abstracting away the different versions of ION uapi, and now `libdmabufheap` abstracts away the DMA-BUF heap uapi while also providing a fallback to `libion` on older kernels.

---

## TODO
- [x] No legacy support for FPC implemented yet (only Egistec), which should be trivial using `BufferAllocatorWrapper.h`;

## Tested on

- [x] PDX215 (Xperia 1III, Sagami, kernel 5.4, non-legacy ION);
- [x] Tama (because it uses the old FPC code that hasn't been ported yet!).

MarijnS95 committed:
Minimal codebase cleanups

MarijnS95 committed:
Refactor local imports to be relative

MarijnS95 committed:
Run newer `clang-format` 15.0.07 over the whole codebase

MarijnS95 committed:
Also run `clang-format` over `c` files

MarijnS95 committed:
Replace `libion` with `libdmabufheap`

Since kernel 5.10 ION has finally been replaced with its DMA-BUF heap
successor, applying lessons learned to the kernel API.  `libion` was
already a compatibility library abstracting away the different versions
of ION uapi, and now `libdmabufheap` abstracts away the DMA-BUF heap
uapi while also providing a fallback to `libion` on older kernels.

jerpelea commented on QSEEComFunc.h:
```diff
@@ -64,8 +64,8 @@ typedef struct qsee_handle_t {
     // Utility functions provided by this library
-    ion_free_def ion_free;
-    ion_alloc_def ion_alloc;
+    // ion_free_def ion_free;
```
remove

jerpelea commented on QSEEComFunc.h:
```diff
@@ -18,7 +18,7 @@
 
 #include "QSEEComAPI.h"
-#include "ion_buffer.h"
+// #include "ion_buffer.h"
```
remove

jerpelea commented on QSEEComFunc.c:
```diff
@@ -72,8 +72,8 @@ int32_t qsee_open_handle(struct qsee_handle_t **ret_handle) {
-    handle->ion_alloc = qcom_km_ion_memalloc;
-    handle->ion_free = qcom_km_ion_dealloc;
+    // handle->ion_alloc = qcom_km_ion_memalloc;
+    // handle->ion_free = qcom_km_ion_dealloc;
```
remove

MarijnS95 replied:
Neither of these can be removed (or we remove all [legacy] FPC support!), since we still have Tama on A13? At least I cannot find it in https://developer.sony.com/develop/open-devices/downloads/software-binaries, maybe we should remove this legacy cruft now!

EDIT: Then we do need a T branch :)

jerpelea replied:
we will do a T branch

MarijnS95 replied:
Since we "decided to keep supporting" Tama on A13 (or at the very least not drop this outright), we just need a branch for Android 12 where `libdmabufheap` was first introduced, if https://source.android.com/docs/core/architecture/kernel/dma-buf-heaps is to be believed. I'll port the legacy code to this library as well.

jerpelea reviewed (approved):
LGTM

MarijnS95 committed:
Clean up typedefs when tagged `struct <structname>` is used

It is typical to either `typdef` a struct - in which case the struct
does not need a name itself - or use the `struct` tag in front of the
struct name.  However, this codebase was mixing and matching both,
giving the struct name the `_t` suffix as well as then using it either
with or without `struct` tag.  Clean this up by at least removing the
`typedef` and `_t` suffix from all structs that are mostly or
exclusively used as tagged structs.

moralrecordings commented:
Just a heads up that this needs a change in https://github.com/sonyxperiadev/local_manifests for Android revisions below 12 to include libdmabufheap, or a tag to pin it to a previous revision. `vendor-sony-oss-fingerprint` is set to pull the master revision, which leads to this build error:
```vendor/oss/fingerprint/Android.mk: error: "android.hardware.biometrics.fingerprint@2.1-service.sony (EXECUTABLES android-arm64) missing libdmabufheap (SHARED_LIBRARIES android-arm64)" ```

(just tried it with android-11.0.0_r46)

MarijnS95 commented:
@moralrecordings it is impossible _and a waste of time_ to port or include `libdmabufheaps` to/on older AOSP, instead as said in https://github.com/sonyxperiadev/vendor-sony-oss-fingerprint/pull/76#discussion_r1214504515 we need an `r-mr1` branch pinned on ea06c50632df7fe7bf311f6f511e5571d30391be and used there. Q already has a similar backport-branch too.
